Fix parameters arithmetic

While trying to parametrise an `AnalogRot` that error is raised:

```python
tests/qadence/test_mitigation.py:134: in <module>
    chain(AnalogRot(duration="t", omega="omega", delta="delta"), AnalogRZ(pi)),
qadence/operations.py:1196: in AnalogRot
    alpha = duration * sympy.sqrt(omega**2 + delta**2) / 1000  # type: ignore [operator]
E   TypeError: unsupported operand type(s) for ** or pow(): 'str' and 'int'
```

--- a/qadence/backends/pulser/pulses.py
+++ b/qadence/backends/pulser/pulses.py
@@ -204,7 +204,7 @@ def analog_rot_pulse(
 
     # create amplitude waveform
     amp_wf = SquareWaveform.from_duration(
-        duration=duration,  # type: ignore
+        duration=abs(duration),  # type: ignore
         max_amp=max_amp,  # type: ignore[arg-type]
         duration_steps=channel.clock_period,  # type: ignore[attr-defined]
         min_duration=channel.min_duration,
@@ -212,7 +212,7 @@ def analog_rot_pulse(
 
     # create detuning waveform
     det_wf = SquareWaveform.from_duration(
-        duration=duration,  # type: ignore
+        duration=abs(duration),  # type: ignore
         max_amp=max_det,  # type: ignore[arg-type]
         duration_steps=channel.clock_period,  # type: ignore[attr-defined]
         min_duration=channel.min_duration,

--- a/qadence/backends/pulser/pulses.py
+++ b/qadence/backends/pulser/pulses.py
@@ -122,7 +122,6 @@ def add_pulses(
         if block.qubit_support.is_global:
             (uuid, duration) = block.parameters.uuid_param("duration")
             t = evaluate(duration) if duration.is_number else sequence.declare_variable(uuid)
-            print(t)
             pulse = Pulse.ConstantPulse(duration=t, amplitude=0, detuning=0, phase=0)
             sequence.add(pulse, GLOBAL_CHANNEL, "wait-for-all")
 

--- a/qadence/backends/pulser/pulses.py
+++ b/qadence/backends/pulser/pulses.py
@@ -122,6 +122,7 @@ def add_pulses(
         if block.qubit_support.is_global:
             (uuid, duration) = block.parameters.uuid_param("duration")
             t = evaluate(duration) if duration.is_number else sequence.declare_variable(uuid)
+            print(t)
             pulse = Pulse.ConstantPulse(duration=t, amplitude=0, detuning=0, phase=0)
             sequence.add(pulse, GLOBAL_CHANNEL, "wait-for-all")
 
@@ -134,12 +135,12 @@ def add_pulses(
 
     elif isinstance(block, ConstantAnalogRotation):
         ps = block.parameters
-        (a_uuid, alpha) = ps.uuid_param("alpha")
+        (t_uuid, duration) = ps.uuid_param("duration")
         (w_uuid, omega) = ps.uuid_param("omega")
         (p_uuid, phase) = ps.uuid_param("phase")
         (d_uuid, detuning) = ps.uuid_param("delta")
 
-        a = evaluate(alpha) if alpha.is_number else sequence.declare_variable(a_uuid)
+        t = evaluate(duration) if duration.is_number else sequence.declare_variable(t_uuid)
         w = evaluate(omega) if omega.is_number else sequence.declare_variable(w_uuid)
         p = evaluate(phase) if phase.is_number else sequence.declare_variable(p_uuid)
         d = evaluate(detuning) if detuning.is_number else sequence.declare_variable(d_uuid)
@@ -148,10 +149,10 @@ def add_pulses(
         block.eigenvalues_generator = block.compute_eigenvalues_generator(block, qc_register)
 
         if block.qubit_support.is_global:
-            pulse = analog_rot_pulse(a, w, p, d, global_channel, config)
+            pulse = analog_rot_pulse(t, w, p, d, global_channel, config)
             sequence.add(pulse, GLOBAL_CHANNEL, protocol="wait-for-all")
         else:
-            pulse = analog_rot_pulse(a, w, p, d, local_channel, config)
+            pulse = analog_rot_pulse(t, w, p, d, local_channel, config)
             sequence.target(qubit_support, LOCAL_CHANNEL)
             sequence.add(pulse, LOCAL_CHANNEL, protocol="wait-for-all")
 
@@ -184,7 +185,7 @@ def add_pulses(
 
 
 def analog_rot_pulse(
-    alpha: TVar | float,
+    duration: TVar | float,
     omega: TVar | float,
     phase: TVar | float,
     detuning: TVar | float,
@@ -201,9 +202,6 @@ def analog_rot_pulse(
     else:
         max_amp = omega
         max_det = detuning
-
-    # get pulse duration in ns
-    duration = 1000 * abs(alpha) / np.sqrt(omega**2 + detuning**2)
 
     # create amplitude waveform
     amp_wf = SquareWaveform.from_duration(

--- a/qadence/operations.py
+++ b/qadence/operations.py
@@ -1152,10 +1152,11 @@ def AnalogRot(
         ConstantAnalogRotation
     """
     q = _cast(QubitSupport, qubit_support)
-    if isinstance(duration, str):
-        duration = Parameter(duration)
-    alpha = duration * sympy.sqrt(Parameter(omega) ** 2 + Parameter(delta) ** 2) / 1000  # type: ignore [operator]
-
+    duration = Parameter(duration)
+    omega = Parameter(omega)
+    delta = Parameter(delta)
+    phase = Parameter(phase)
+    alpha = duration * sympy.sqrt(omega**2 + delta**2) / 1000
     ps = ParamMap(alpha=alpha, duration=duration, omega=omega, delta=delta, phase=phase)
     return ConstantAnalogRotation(parameters=ps, qubit_support=q)
 

--- a/qadence/operations.py
+++ b/qadence/operations.py
@@ -1154,7 +1154,7 @@ def AnalogRot(
     q = _cast(QubitSupport, qubit_support)
     if isinstance(duration, str):
         duration = Parameter(duration)
-    alpha = duration * sympy.sqrt(Parameter(omega)**2 + Parameter(delta)**2) / 1000  # type: ignore [operator]
+    alpha = duration * sympy.sqrt(Parameter(omega) ** 2 + Parameter(delta) ** 2) / 1000  # type: ignore [operator]
 
     ps = ParamMap(alpha=alpha, duration=duration, omega=omega, delta=delta, phase=phase)
     return ConstantAnalogRotation(parameters=ps, qubit_support=q)

--- a/qadence/operations.py
+++ b/qadence/operations.py
@@ -1154,7 +1154,7 @@ def AnalogRot(
     q = _cast(QubitSupport, qubit_support)
     if isinstance(duration, str):
         duration = Parameter(duration)
-    alpha = duration * sympy.sqrt(omega**2 + delta**2) / 1000  # type: ignore [operator]
+    alpha = duration * sympy.sqrt(Parameter(omega)**2 + Parameter(delta)**2) / 1000  # type: ignore [operator]
 
     ps = ParamMap(alpha=alpha, duration=duration, omega=omega, delta=delta, phase=phase)
     return ConstantAnalogRotation(parameters=ps, qubit_support=q)

--- a/tests/analog/test_pyq_vs_pulser.py
+++ b/tests/analog/test_pyq_vs_pulser.py
@@ -5,6 +5,7 @@ import pytest
 import torch
 from metrics import ATOL_DICT, JS_ACCEPTANCE, LARGE_SPACING, SMALL_SPACING
 
+from qadence import Parameter
 from qadence.analog import RealisticDevice, RydbergDevice
 from qadence.blocks import AbstractBlock, chain, kron
 from qadence.circuit import QuantumCircuit
@@ -31,6 +32,58 @@ from qadence.parameters import FeatureParameter
 from qadence.register import Register
 from qadence.states import equivalent_state, random_state
 from qadence.types import BackendName, DiffMode
+
+
+@pytest.mark.parametrize(
+    "duration, omega, delta, phase, values",
+    [
+        (
+            "t",
+            "omega",
+            "delta",
+            "phase",
+            {
+                "t": torch.tensor([1000.0]),
+                "omega": torch.tensor([1.0]),
+                "delta": torch.tensor([1.0]),
+            },
+        ),
+        (
+            Parameter("t"),
+            Parameter("omega"),
+            Parameter("delta"),
+            Parameter("phase"),
+            {
+                "t": torch.tensor([1000.0]),
+                "omega": torch.tensor([1.0]),
+                "delta": torch.tensor([1.0]),
+            },
+        ),
+        (
+            1000.0,
+            1.0,
+            1.0,
+            1.0,
+            {},
+        ),
+    ],
+)
+def test_parametrized_analog_rot(
+    duration: float | str | Parameter,
+    omega: float | str | Parameter,
+    delta: float | str | Parameter,
+    phase: float | str | Parameter,
+    values: dict,
+) -> None:
+    analog_rot = AnalogRot(duration=duration, omega=omega, delta=delta)
+
+    register = Register.line(2, spacing=8)
+    circuit = QuantumCircuit(register, analog_rot)
+
+    wf_pyq = run(circuit, values=values, backend=BackendName.PYQTORCH)
+    wf_pulser = run(circuit, values=values, backend=BackendName.PULSER)
+
+    assert equivalent_state(wf_pyq, wf_pulser, atol=ATOL_DICT[BackendName.PULSER])
 
 
 @pytest.mark.flaky(max_runs=5)

--- a/tests/backends/pulser_basic/test_rotations.py
+++ b/tests/backends/pulser_basic/test_rotations.py
@@ -45,7 +45,7 @@ def test_single_rotation(block: AbstractBlock, goal: Counter) -> None:
 def test_single_rotation_multiple_qubits(
     single_rotation: AbstractBlock, global_rotation: AbstractBlock
 ) -> None:
-    register = Register.from_coordinates([(-0.5, 0), (0.5, 0)], lattice="line")
+    register = Register.from_coordinates([(-0.5, 0), (0.5, 0)], lattice="line", spacing=8.0)
 
     circuit1 = QuantumCircuit(register, single_rotation)
     model_pulser1 = QuantumModel(

